src/Swarm/Game/Step.hs

```diff
@@ -803,6 +803,64 @@ execConst c vs k = do
     Raise -> case vs of
       [VString s] -> return $ Up (User s) k
       _ -> badConst
+    Reprogram -> case vs of
+      [VString otherRobotName, VDelay _ cmd e] -> do
+        em <- lift . lift $ use entityMap
+        mode <- lift . lift $ use gameMode
+
+        -- check if robot exists
+        otherRobot <-
+          robotNamed otherRobotName
+            >>= (`isJustOr` cmdExn Reprogram ["There is no robot named", otherRobotName, "."])
+
+        -- check if robot has completed executing it's current command
+        _ <-
+          finalValue (otherRobot ^. machine)
+            `isJustOr` cmdExn
+              Reprogram
+              ["You cannot reprogram a robot that has not completed it's current command"]
+
+        let -- Standard devices that are always installed.
+            -- XXX in the future, make a way to build these and just start the base
+            -- out with a large supply of each?
+            stdDeviceList = ["treads", "grabber", "solar panel", "detonator", "scanner"]
+            stdDevices = S.fromList $ mapMaybe (`lookupEntityName` em) stdDeviceList
+
+            -- Find out what capabilities are required by the program that will
+            -- be run on the other robot, and what devices would provide those
+            -- capabilities.
+            (caps, _capCtx) = requiredCaps (snd (otherRobot ^. robotCtx)) cmd
```

Comment:
I don't think it makes sense to use the `otherRobot`'s `robotCtx` to check this, especially if we are later going to reset it to empty.  Honestly I am not quite sure what the right thing to do is, I will have to think about it a bit...

Reply:
I think we can perform this check against an empty set, if we assume that reprogram requires all devices to already be preinstalled on the target robot, and that reprogrammed robot is starting it's computation afresh.

Reply:
I'm not sure that would be correct.  The context given as an argument to `requiredCaps` is supposed to map names of definitions to the capabilities required by those definitions.  If we pass the empty context to `requiredCaps` then it might return fewer capabilities than it should, if the given command refers to some definitions in the environment.  That would lead to a situation where the `reprogram` command is successful and the reprogrammed robot starts running its program, but later crashes due to not having the right capabilities to execute something.  In that situation we would rather have it fail right away (simply as a kindness to the user).

The environment used for the reprogrammed robot (the `e` on line 862, `In cmd e [FExec]`) came from the `VDelay` value on line 807 (`VDelay _ cmd e`), which in turn came from the environment active in the parent robot when the `delay` was processed.  That means the reprogrammed robot will now know about all the definitions the parent robot knew about (which is exactly as it should be).  That means the `capCtx` we should be passing to `requiredCaps` here is the one stored in the *parent* robot's `robotCtx` (i.e. `rctx@(_, capCtx) <- use robotCtx`).

This also means the `robotCtx` of the reprogrammed robot should be set to `rctx`, the same context as its parent (in case it, in turn, reprograms another robot, etc.).

It probably doesn't matter as much what the `robotEnv` is set to, since that is used more for the `base` and its ability to make new definitions, but on principle we might as well just set it to the same thing as its parent robot.

Reply:
Probably this also means `robotEnv` is a bad name!  I'm open to better suggestions.